src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -7743,11 +7762,16 @@ private ForStatementSyntax ParseForStatement()
                 }
                 else
                 {
-                    isDeclaration = !this.IsQueryExpression(mayBeVariableDeclaration: true, mayBeMemberDeclaration: false) &&
-                                    this.ScanType() != ScanTypeFlags.NotType &&
-                                    this.IsTrueIdentifier();
+                    decl = TryParseDeconstructionDeclaration(withEquals: true);
 
-                    this.Reset(ref resetPoint);
+                    if (decl == null)
+                    {
+                        isDeclaration = !this.IsQueryExpression(mayBeVariableDeclaration: true, mayBeMemberDeclaration: false) &&
+                                        this.ScanType() != ScanTypeFlags.NotType &&
+                                        this.IsTrueIdentifier();
+
+                        this.Reset(ref resetPoint);
+                    }
```

Comment:
@AlekseyTs 
Thanks for catching that. I didn't intend to execute the block below.
I will add a condition check (don't execute if we have a deconstruction).
```C#
else if (decl == null) 
{
    // Not a type followed by an identifier, so it must be an expression list.
    if (this.CurrentToken.Kind != SyntaxKind.SemicolonToken)
    {
        this.ParseForStatementExpressionList(ref openParen, initializers);
    }
}
```